librz/include/rz_types_overflow.h

```diff
@@ -8,55 +8,78 @@
 // ADD
 // if ((x > 0) && (a > INT_MAX - x)) /* `a + x` would overflow */;
 // if ((x < 0) && (a < INT_MIN - x)) /* `a + x` would underflow */;
-#define SZT_ADD_OVFCHK(x, y)  ((SIZE_MAX - (x)) < (y))
-#define SSZT_ADD_OVFCHK(a, x) ((((x) > 0) && ((a) > SSIZE_MAX - (x))) || (((x) < 0) && (a) < SSIZE_MIN - (x)))
-#define UT64_ADD_OVFCHK(x, y) ((UT64_MAX - (x)) < (y))
-#define ST64_ADD_OVFCHK(a, x) ((((x) > 0) && ((a) > ST64_MAX - (x))) || (((x) < 0) && (a) < ST64_MIN - (x)))
-#define UT32_ADD_OVFCHK(x, y) ((UT32_MAX - (x)) < (y))
-#define ST32_ADD_OVFCHK(a, x) ((((x) > 0) && ((a) > ST32_MAX - (x))) || (((x) < 0) && (a) < ST32_MIN - (x)))
-#define UT16_ADD_OVFCHK(x, y) ((UT16_MAX - (x)) < (y))
-#define ST16_ADD_OVFCHK(a, b) ((((b) > 0) && ((a) > ST16_MAX - (b))) || (((b) < 0) && ((a) < ST16_MIN - (b))))
-#define UT8_ADD_OVFCHK(x, y)  ((UT8_MAX - (x)) < (y))
-#define ST8_ADD_OVFCHK(a, x)  ((((x) > 0) && ((a) > ST8_MAX - (x))) || ((x) < 0 && (a) < ST8_MIN - (x)))
+
+// #define SZT_ADD_OVFCHK(x, y)  ((SIZE_MAX - (x)) < (y))
+// #define SSZT_ADD_OVFCHK(a, x) ((((x) > 0) && ((a) > SSIZE_MAX - (x))) || (((x) < 0) && (a) < SSIZE_MIN - (x)))
+// #define UT64_ADD_OVFCHK(x, y) ((UT64_MAX - (x)) < (y))
+// #define ST64_ADD_OVFCHK(a, x) ((((x) > 0) && ((a) > ST64_MAX - (x))) || (((x) < 0) && (a) < ST64_MIN - (x)))
+// #define UT32_ADD_OVFCHK(x, y) ((UT32_MAX - (x)) < (y))
+// #define ST32_ADD_OVFCHK(a, x) ((((x) > 0) && ((a) > ST32_MAX - (x))) || (((x) < 0) && (a) < ST32_MIN - (x)))
+// #define UT16_ADD_OVFCHK(x, y) ((UT16_MAX - (x)) < (y))
+// #define ST16_ADD_OVFCHK(a, b) ((((b) > 0) && ((a) > ST16_MAX - (b))) || (((b) < 0) && ((a) < ST16_MIN - (b))))
+// #define UT8_ADD_OVFCHK(x, y)  ((UT8_MAX - (x)) < (y))
+// #define ST8_ADD_OVFCHK(a, x)  ((((x) > 0) && ((a) > ST8_MAX - (x))) || ((x) < 0 && (a) < ST8_MIN - (x)))
+
+
+//Using CLANG/GCC builtins
+
+#define SZT_ADD_OVFCHK(x, y)  __builtin_add_overflow_p (x, y, (__typeof__ ((x) + (y))) 0)
+#define SZT_ADD_OVFCHK(x, y)  __builtin_add_overflow_p (x, y, (__typeof__ ((x) + (y))) 0)
+#define SSZT_ADD_OVFCHK(a, x) __builtin_add_overflow_p (a, x, (__typeof__ ((a) + (x))) 0)
+#define UT64_ADD_OVFCHK(x, y) __builtin_add_overflow_p (x, y, (__typeof__ ((x) + (y))) 0)
+#define ST64_ADD_OVFCHK(a, x) __builtin_add_overflow_p (a, x, (__typeof__ ((a) + (x))) 0)
+#define UT32_ADD_OVFCHK(x, y) __builtin_add_overflow_p (x, y, (__typeof__ ((x) + (y))) 0)
+#define ST32_ADD_OVFCHK(a, x) __builtin_add_overflow_p (a, x, (__typeof__ ((a) + (x))) 0)
+#define UT16_ADD_OVFCHK(x, y) __builtin_add_overflow_p (x, y, (__typeof__ ((x) + (y))) 0)
+#define ST16_ADD_OVFCHK(a, b) __builtin_add_overflow_p (a, b, (__typeof__ ((a) + (b))) 0)
+#define UT8_ADD_OVFCHK(x, y)  __builtin_add_overflow_p (x, y, (__typeof__ ((x) + (y))) 0)
+#define ST8_ADD_OVFCHK(a, x)  __builtin_add_overflow_p (a, x, (__typeof__ ((a) + (x))) 0)
```

Comment:
these are not compatible with MSVC

Reply:
if you want you can make an `#if __clang__ || __gcc__` and have the else with the old check

Reply:
No,  they should be checked in meson, just how it was done for bswap builtins, I referred corresponding commit in the issue

Reply:
makes sense.

Reply:
I have made following changes.
1. In the Meson build file (meson.build), I've added checks for the availability of these built-in functions.
![Screenshot from 2024-03-16 15-38-33](https://github.com/rizinorg/rizin/assets/94524217/4da89757-e420-4928-8ae0-a46fb0a45fa2)

2.In the rz_types.h header file, I've defined macros HAVE___BUILTIN_ADD_OVERFLOW, HAVE___BUILTIN_SUB_OVERFLOW, and HAVE___BUILTIN_MUL_OVERFLOW based on the values obtained from Meson.
![Screenshot from 2024-03-16 15-40-30](https://github.com/rizinorg/rizin/assets/94524217/e5c764bb-39f5-4af1-8082-6a4ae3dc816a)

3. I've used conditional compilation (#ifdef) in the rz_types_overflow.h to define macros depending on whether these built-in functions are available.
![Screenshot from 2024-03-16 15-39-59](https://github.com/rizinorg/rizin/assets/94524217/113f2350-f26b-4bc6-8254-655bc03f5358)

Could you please review these changes and confirm whether I'm on the right track? If there are any additional steps I should take or if any adjustments are needed, please let me know.

Reply:
Please avoid asking via images - it makes reviewing harder. Just push whatever changes directly to the PR, if any changes will be necessary, we will request them via the review process.

Reply:
For defines, please use `rz_userconf.h`, it is autogenerated based on the meson detection, see how it's done in the commit I referenced.